### PR TITLE
Stylegan branch 

### PR DESCRIPTION
**Pull request by grenlayk**

### CODE REVIEW

File: src/training/stylegan.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        self.device = device
+        self.model = model
+        self.optimizer = optimizer
+        self.scheduler = scheduler
+        self.train_dataloader = train_dataloader
+        self.val_dataloader = val_dataloader
+        self.total_epochs = total_epochs
+        self.logger = logger
+        self.storage = storage
+        self.ocr_loss = ocr.OCRLoss().to(device)
```
Comment (grenlayk) [inline src/training/stylegan.py:33]: Maybe I wrong, but I suppose all initialisation should happen in config and only result passed to the trainer class.

File: src/training/stylegan.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        self.perceptual_loss = perceptual_loss.to(device)
+        self.coef_ocr = coef_ocr_loss
+        self.coef_perceptual = coef_perceptual_loss
+        model_ft = models.resnet18(pretrained=True)
```
Comment (grenlayk) [inline src/training/stylegan.py:37]: Same goes here.

File: src/training/stylegan.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            {'model': self.model, 'optimizer': self.optimizer, 'scheduler': self.scheduler},
+            avg_losses['full_loss']
+        )
+
+
+    def run(self):
+        for epoch in range(self.total_epochs):
+            self.train()
+            with torch.no_grad():
+              if (epoch + 3) % 10 == 0:
```
Comment (grenlayk) [inline src/training/stylegan.py:97]: If you are doing this because validation takes too long, maybe you could add `cut` options to `BaselineDataset` as we did it in `ColorDataset`: [link](https://github.com/grenlayk/text-deep-fake/blob/dc0be62766452808f141ca039c89b880aa4e833d/src/data/color.py#L17)

File: src/config/stylegan.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        x = self.maxpool(x)
+
+        x = self.layer1(x)
+        x = self.layer2(x)
+        x = self.layer3(x)
+        x = self.layer4(x)
+
+        #x = self.avgpool(x)
+        #x = torch.flatten(x, 1)
+        #x = self.fc(x)
```
Comment (SphericalPotatoInVacuum) [inline src/config/stylegan.py:32]: In a perfect world these should have a note why they were commented out or deleted altogether, but I'm not saying that I'm not nitpicking right now.

File: src/config/stylegan.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        return x
+
+class Config:
+    def __init__(self):
+        disk.login()
+
+        device = 'cuda:0' if torch.cuda.is_available() else 'cpu'
+        info_logger.info(f'Using device: {device}')
+        data_dir = Path("data/imgur")
+        style_dir = data_dir / 'IMGUR5K_small'
```
Comment (SphericalPotatoInVacuum) [inline src/config/stylegan.py:43]: Correct dataset is IMGUR5K_REDUCED, it has fixed json mapping thingy.

File: src/config/stylegan.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        if not data_dir.exists():
+            data_dir.mkdir()
+            local_path = download_data(Path("data/IMGUR5K_small.tar"), data_dir)
+            unarchieve(local_path)
+        batch_size = 16
+        train_dataloader = DataLoader(BaselineDataset(style_dir / 'train'), shuffle=True, batch_size = batch_size)
+        val_dataloader = DataLoader(BaselineDataset(style_dir / 'val'), batch_size = batch_size)
+
+        total_epochs = 500 #20
```
Comment (SphericalPotatoInVacuum) [inline src/config/stylegan.py:52]: My #20 comment is not needed anymore :)

File: src/config/stylegan.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        model = StyleBased_Generator(dim_latent=512)
+        #model.load_state_dict(torch.load('/content/text-deep-fake/checkpoints/stylegan_one_style_working/14/model'))
```
Comment (SphericalPotatoInVacuum) [inline src/config/stylegan.py:54]: This one too can go I believe

File: src/config/stylegan.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        model.to(device)
+        style_embedder = models.resnet18()
+        style_embedder.fc = torch.nn.Identity()
+        style_embedder = style_embedder.to(device)
+        content_embedder = ContentResnet(BasicBlock, [2, 2, 2, 2]).to(device)
```
Comment (SphericalPotatoInVacuum) [inline src/config/stylegan.py:59]: I suggest moving them inside the model

File: src/losses/typeface_perceptual.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
+import torch
+import torchvision
+from src.disk import disk 
+from pathlib import Path
+class TypefacePerceptualLoss(torch.nn.Module):
+    def __init__(self, model_remote_path = 'checkpoints/typeface/2/model', model_local_path = 'typeface', resize=True):
+        super(TypefacePerceptualLoss, self).__init__()
```
Comment (SphericalPotatoInVacuum) [inline src/losses/typeface_perceptual.py:7]: `super().__init__()`

File: src/training/stylegan.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        self.scheduler = scheduler
+        self.train_dataloader = train_dataloader
+        self.val_dataloader = val_dataloader
+        self.total_epochs = total_epochs
+        self.logger = logger
+        self.storage = storage
+        self.ocr_loss = ocr_loss.to(device)
+        self.perceptual_loss = perceptual_loss.to(device)
+        self.coef_ocr = coef_ocr_loss
+        self.coef_perceptual = coef_perceptual_loss
```
Comment (SphericalPotatoInVacuum) [inline src/training/stylegan.py:40]: We have `ComposeLoss` so we could pass it as one parameter

**Review by SphericalPotatoInVacuum — COMMENTED** [review]
Looks ok apart from some minute details, but nothing too serious, almost all of it can wait for refactoring step (sorry whoever reads this afterwards for encouraging bad practices but we're on a deadline :smiling_face_with_tear:)

**Review by SphericalPotatoInVacuum — APPROVED** [review]

File: src/config/stylegan.py (hunk 4 above)
Comment (grenlayk) [reply src/config/stylegan.py:33]: I think it's alright how it is, because it's clear how did we modify this block.

File: src/config/stylegan.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        total_epochs = 500
+
+        model = StyleBased_Generator(dim_latent=512)
+        model.load_state_dict(torch.load('models/Stylegan_content_192x64/model'))
+        model.to(device)
+
+        style_embedder = models.resnet18()
+        style_embedder.fc = torch.nn.Identity()
+        style_embedder = style_embedder.to(device)
+        style_embedder.load_state_dict(torch.load('models/Stylegan_content/style_embedder'))
```
Comment (grenlayk) [inline src/config/stylegan.py:59]: I suppose there should be `Stylegan_content_192x64` instead of `Stylegan_content`

File: src/config/stylegan.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        storage = Storage('checkpoints/stylegan(pretrained_on_content)_gram_ocr_192x64')
+
+        logger = Logger(
+            image_freq=100,
+            project_name='StyleGan',
+            tags=('gram', 'pretrained_on_content', 'trba_ocr', 'full_dataset'),
+            config={
+                'ocr_coef': ocr_coef,
+                'perceptual_coef': perceptual_coef,
+                'style_layers': [2, 3],
```
Comment (grenlayk) [inline src/config/stylegan.py:90]: no cycle and recon coefs here

File: src/config/stylegan.py (hunk 11 above)
Comment (grenlayk) [reply src/config/stylegan.py:59]: also I suppose we can have something similar to  `weights_folder = 'models/Stylegan_content_192x64'` and then use it like this:
```Python
style_embedder.load_state_dict(torch.load(f'{weights_folder}/style_embedder')) 
```

File: src/data/baseline.py [hunk 13]
```diff
@@ -64,13 +63,25 @@ def __getitem__(self, index):
...
+            img_content_style = np.array(pil_content_style)
+            img_content_style = cv2.resize(img_content_style, img_size)
+
+            img_content_style = img_content_style * 1.0 / 255
+            img_content_style = torch.from_numpy(np.transpose(img_content_style[:, :, [2, 1, 0]], (2, 0, 1))).float()
+
+            return img_style, img_content, content, img_content_style
 
         except Exception as e:
+            raise e
```
Comment (grenlayk) [inline src/data/baseline.py:85]: Let's have it after logger message maybe? Also in this way lines after are not needed.

File: src/losses/perceptual.py [hunk 14]
```diff
@@ -18,7 +18,7 @@ def __init__(self, resize=False):
         self.register_buffer("mean", torch.tensor([0.485, 0.456, 0.406]).view(1, 3, 1, 1))
         self.register_buffer("std", torch.tensor([0.229, 0.224, 0.225]).view(1, 3, 1, 1))
 
-    def forward(self, input, target, feature_layers=[0, 1, 2, 3], style_layers=[]):
+    def forward(self, input, target, feature_layers=[0, 1, 2, 3], style_layers=[1,2]):
```
Comment (grenlayk) [inline src/losses/perceptual.py:21]: It's not the best idea to use `style_layers` here because gram loss is not normalised here. Check Gram loss for reference: https://github.com/grenlayk/text-deep-fake/blob/53f9d2feb0752845c8a99ae0c1f99e679e207943/src/losses/gram.py#L44

File: src/losses/typeface_perceptual.py (hunk 9 above)
Comment (grenlayk) [reply src/losses/typeface_perceptual.py:7]: I think we can leave it like this for now and you will change it with other perceptual losses later

File: src/training/stylegan.py (hunk 10 above)
Comment (grenlayk) [reply src/training/stylegan.py:44]: From one side I agree, from the other -- it's more clear and we also have different objects to pass to these losses, so let's leave them separately, we won't suffer from 2 extra lines.

File: src/training/stylegan.py [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+                    'content': content_batch,
+                    'result': res,
+                    'recognized': word_images})
+
+    def validate(self, epoch):
+        self.model.eval()
+        self.content_embedder.eval()
+        self.style_embedder.eval()
+
+        for style_batch, content_batch, label_batch in self.val_dataloader:
```
Comment (grenlayk) [inline src/training/stylegan.py:116]: Are you sure there shouldn't be 4 args ?

File: src/training/stylegan.py [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+                           'optimizer': self.optimizer,
+                           'scheduler': self.scheduler},
+                          avg_losses['full_loss'])
+
+    def run(self):
+        for epoch in range(self.total_epochs):
+            self.train()
+            with torch.no_grad():
+                self.validate(epoch)
+            if self.scheduler is not None:
```
Comment (grenlayk) [inline src/training/stylegan.py:153]: I am still puzzled -- why do we need it? Why do we do step here?

File: src/training/stylegan.py (hunk 10 above)
Comment (SphericalPotatoInVacuum) [reply src/training/stylegan.py:44]: True, I forgot that we calculate loss between different objects here, no objections here anymore

File: src/training/stylegan.py (hunk 16 above)
Comment (SphericalPotatoInVacuum) [reply src/training/stylegan.py:153]: Usually a scheduler step is done each epoch

Comment (nikoryagin) [reply src/training/stylegan.py:153]: what is the question? didnt get it
